default.nix

```diff
@@ -30,63 +30,71 @@ let
     nix-store --init
   '';
 
-  build = pkgs.callPackage ./package.nix {
-    inherit nixpkgsLibPath initNix runtimeExprPath testNixpkgsPath;
-  };
 
-in
-build // {
+  results = {
+    build = pkgs.callPackage ./package.nix {
+      inherit nixpkgsLibPath initNix runtimeExprPath testNixpkgsPath;
+    };
 
-  inherit build;
+    shell = pkgs.mkShell {
+      env.NIX_CHECK_BY_NAME_EXPR_PATH = toString runtimeExprPath;
+      env.NIX_PATH = "test-nixpkgs=${toString testNixpkgsPath}:test-nixpkgs/lib=${toString nixpkgsLibPath}";
+      env.RUST_SRC_PATH = "${pkgs.rustPlatform.rustLibSrc}";
+      inputsFrom = [ results.build ];
+      nativeBuildInputs = with pkgs; [
+        npins
+        rust-analyzer
+      ];
+    };
 
-  # Used by CI and good for debugging
-  inherit pkgs;
+    # Run regularly by CI and turned into a PR
+    autoPrUpdate = pkgs.writeShellApplication {
+      name = "auto-pr-update";
+      runtimeInputs = with pkgs; [
+        npins
+        cargo
+      ];
+      text =
+        let
+          commands = {
+            "npins changes" = ''
+              npins update --directory "$REPO_ROOT/npins"'';
+            "cargo changes" = ''
+              cargo update --manifest-path "$REPO_ROOT/Cargo.toml"'';
+          };
+        in
+        ''
+          REPO_ROOT=$1
+          echo "Run automated updates"
+        ''
+        + pkgs.lib.concatStrings (pkgs.lib.mapAttrsToList (title: command: ''
+          echo -e '<details><summary>${title}</summary>\n\n```'
+          ${command} 2>&1
+          echo -e '```\n</details>'
+        '') commands);
+    };
 
-  shell = pkgs.mkShell {
-    env.NIX_CHECK_BY_NAME_EXPR_PATH = toString runtimeExprPath;
-    env.NIX_PATH = "test-nixpkgs=${toString testNixpkgsPath}:test-nixpkgs/lib=${toString nixpkgsLibPath}";
-    env.RUST_SRC_PATH = "${pkgs.rustPlatform.rustLibSrc}";
-    inputsFrom = [ build ];
-    nativeBuildInputs = with pkgs; [
-      npins
-      rust-analyzer
-    ];
+    # Tests the tool on the pinned Nixpkgs tree, this is a good sanity check
+    nixpkgsCheck = pkgs.runCommand "test-nixpkgs-check-by-name" {
+      nativeBuildInputs = [
+        results.build
+        pkgs.nix
+      ];
+      nixpkgsPath = nixpkgs;
+    } ''
+      ${initNix}
+      nixpkgs-check-by-name --base "$nixpkgsPath" "$nixpkgsPath"
+      touch $out
+    '';
   };
 
-  # Run regularly by CI and turned into a PR
-  autoPrUpdate = pkgs.writeShellApplication {
-    name = "auto-pr-update";
-    runtimeInputs = with pkgs; [
-      npins
-      cargo
-    ];
-    text =
-      let
-        commands = [
-          "npins update"
-          "cargo update"
-        ];
-      in
-      ''
-        echo "Run automated updates"
-      ''
-      + pkgs.lib.concatMapStrings (command: ''
-        echo -e '<details><summary>${command}</summary>\n\n```'
-        ${command} 2>&1
-        echo -e '```\n</details>'
-      '') commands;
-  };
+in
+results.build // results // {
```

Comment:
why do we need `results.build` to be first?

Reply:
Shouldn't matter, they need to be non-overlapping. In the future I'd like to use `lib.derivations.lazyDerivation` (to avoid `results.build` having so many attributes) and `lib.attrsets.unionOfDisjoint` (to ensure they don't overlap) to make this nicer.